### PR TITLE
CMake travis checks

### PR DESCRIPTION
**Pull request by jandrej**
This adds the cmake builds and unit tests build with cmake to travis.

Please move the check wherever you want in the travis config file. I just put it at the top for the development workflow.

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge| 
 | --- | --- | --- | --- | --- | --- | --- | 
| [#1266](https://github.com/mfem/mfem/pull/1266) | @jandrej | @tzanio | @adrienbernede + @cjvogl | 02/22/20 | 03/16/20 | 03/17/20 | |

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -3,6 +3,7 @@ language: cpp
 sudo: false
 
 stages:
+  - cmake
```
Comment (tzanio) [inline .travis.yml]: Should this be a stage after `checks` or maybe part of it or part of the `tests` stage?

Comment (v-dobrev) [reply .travis.yml]: To me it makes sense to make it a part of the `tests` stage.

Comment (jandrej) [reply .travis.yml]: I'm fine with either. Should I move it to tests?

Comment (tzanio) [reply .travis.yml]: > Should I move it to `tests`?

Yes

File: .travis.yml [hunk 2]
```diff
@@ -166,6 +166,33 @@ jobs:
...
+      addons:
+        apt:
+          packages:
+            - mpich
+            - libmpich-dev
+      env: MPI=YES
+      script:
+        - cd ${TRAVIS_BUILD_DIR}
+        - mkdir ${TRAVIS_BUILD_DIR}/build
+        - cd ${TRAVIS_BUILD_DIR}/build
```
Comment (adrienbernede) [inline .travis.yml]: Why not just:
```suggestion
        - mkdir build && cd build
```

File: .travis.yml [hunk 3]
```diff
@@ -166,6 +166,33 @@ jobs:
...
+        - ls ${TRAVIS_BUILD_DIR}/../metis-4.0
+        - cmake .. -DMFEM_USE_MPI=ON -DHYPRE_DIR=${TRAVIS_BUILD_DIR}/../hypre-2.10.0b/src/hypre -DMETIS_INCLUDE_DIRS=${TRAVIS_BUILD_DIR}/../metis-4.0/Lib -DMETIS_LIBRARIES=${TRAVIS_BUILD_DIR}/../metis-4.0/libmetis.a
+        - make -j3 mfem examples unit_tests
+        - cd ${TRAVIS_BUILD_DIR}/build/tests/unit
```
Comment (adrienbernede) [inline .travis.yml]: Why not just:
```suggestion
        - cd tests/unit
```

**Comment by adrienbernede** [comment]
Well, clearly, I messed up...

File: .travis.yml [hunk 4]
```diff
@@ -166,6 +166,33 @@ jobs:
...
+          packages:
+            - mpich
+            - libmpich-dev
+      env: MPI=YES
+      script:
+        - cd ${TRAVIS_BUILD_DIR}
+        - mkdir ${TRAVIS_BUILD_DIR}/build
+        - cd ${TRAVIS_BUILD_DIR}/build
+        - ls ${TRAVIS_BUILD_DIR}/../metis-4.0
+        - cmake .. -DMFEM_USE_MPI=ON -DHYPRE_DIR=${TRAVIS_BUILD_DIR}/../hypre-2.10.0b/src/hypre -DMETIS_INCLUDE_DIRS=${TRAVIS_BUILD_DIR}/../metis-4.0/Lib -DMETIS_LIBRARIES=${TRAVIS_BUILD_DIR}/../metis-4.0/libmetis.a
```
Comment (adrienbernede) [inline .travis.yml]: DMETIS_**INCLUDE**_DIRS=${TRAVIS_BUILD_DIR}/../metis-4.0/**Lib**
Is that really the case?

Comment (v-dobrev) [reply .travis.yml]: That's where the headers are with METIS 4. However, we do not actually include any of those headers in MFEM -- we just copied the prototypes of the functions we use, e.g. in `mesh.cpp`. We only really need the METIS include path when using METIS 5.

On the other hand, in the CMake build system we (probably) expect `*_INCLUDE_DIRS` to be defined for every library that is enabled.

File: .travis.yml [hunk 5]
```diff
@@ -166,6 +166,33 @@ jobs:
...
+        apt:
+          packages:
+            - mpich
+            - libmpich-dev
+      env: MPI=YES
+      script:
+        - cd ${TRAVIS_BUILD_DIR}
+        - mkdir ${TRAVIS_BUILD_DIR}/build
+        - cd ${TRAVIS_BUILD_DIR}/build
+        - ls ${TRAVIS_BUILD_DIR}/../metis-4.0
```
Comment (cjvogl) [inline .travis.yml]: Full disclosure: I've never worked with anything but a very simple Travis script.  Is this `ls` necessary/intended here or is it leftover debugging code?

Comment (jandrej) [reply .travis.yml]: This looks like a debug leftover, yes.

**Comment by tzanio** [comment]
@adrienbernede  and @cjvogl, have all your questions been answered?

File: .travis.yml (hunk 5 above)
Comment (cjvogl) [reply .travis.yml]: Once it is removed, I think this should be merged.

**Comment by cjvogl** [comment]
> 
> 
> @adrienbernede and @cjvogl, have all your questions been answered?

My question has been answered, but it has prompted a requested change (that the code that is indeed leftover debug code be removed before merging this PR).

**Comment by jandrej** [comment]
@cjvogl removed the `ls` command.

File: .travis.yml (hunk 4 above)
Comment (adrienbernede) [reply .travis.yml]: @jandrej @v-dobrev 
After further investigation, both `METIS_INCLUDE_DIRS` and `METIS_LIBRARIES` are unnecessary here. In fact `metis-4.0` is at the default location, so `FindMetis` will find it and configure it.

I like to see CI builds as a demonstrator of "how to build MFEM" though. So I would agreed with specifying `METIS_DIR`, which seems the most appropriate way to manually specify a location for metis in this project. 

Does it sound reasonable?

To sum-up, I suggest either:
- completely removing `-DMETIS_INCLUDE_DIRS=${TRAVIS_BUILD_DIR}/../metis-4.0/Lib -DMETIS_LIBRARIES=${TRAVIS_BUILD_DIR}/../metis-4.0/libmetis.a`
- or replacing them with `-DMETIS_DIR=${TRAVIS_BUILD_DIR}/../metis-4.0`

Comment (v-dobrev) [reply .travis.yml]: I'm fine with both choices, with a slight preference for the second one.

**Comment by v-dobrev** [comment]
It looks like we need to revert the full `METIS` options (`METIS_INCLUDE_DIRS` and `METIS_LIBRARIES`) -- autodetect does not work because we only cache the library file `libmetis.a` but not the headers.

**Comment by adrienbernede** [comment]
@v-dobrev why not caching the header files just like in a regular installation?

I know we don’t _need_ them but having to manually specify a fake `METIS_INCLUDE_DIRS` is not a better option IMO.

**Comment by adrienbernede** [comment]
Solved in #1345 but running into unstable travis failures...

**Comment by v-dobrev** [comment]
@adrienbernede, #1345 look good to me too.

**Comment by v-dobrev** [comment]
I think I need to purge the cache and restart the build on travis. I'll do that in a moment.

**Comment by v-dobrev** [comment]
The cmake test on travis worked fine. I'll restart one more time to make sure it works when using the cache.

**Comment by v-dobrev** [comment]
The tests look good.

As we discussed with @jandrej, I'll suggest one small change: instead of running the unit tests directly with `./unit_tests`, use `ctest --output-on-failure`. This may require running `make -j3` in the `tests/unit` directory before running `ctest` to make sure all test in that directory are built.

**Comment by v-dobrev** [comment]
Hmm it looks like the CMake build on travis stalled -- I'll restart that test to see if this is a real issue.

**Comment by adrienbernede** [comment]
@v-dobrev @jandrej to be honest, the tests seem stall prone since my change in the caching. I just don’t know how to explain that.

**Comment by v-dobrev** [comment]
> @v-dobrev @jandrej to be honest, the tests seem stall prone since my change in the caching. I just don’t know how to explain that.

@adrienbernede, the tests were fine after your update -- I just had to delete the old cache -- see my comments above from 2 days ago. You can also see the green check mark next to your commit.

The new problem is still there after one restart, so we'll need to look into it. It looks like the test `psedov_tests_cpu_np=4` stalls for some reason.

**Comment by jandrej** [comment]
Maybe because travis only allows 3 ranks?

**Comment by v-dobrev** [comment]
> Maybe because travis only allows 3 ranks?

This is not an issue in the GNU make build where we also use 4 processors.

To debug, one thing we can try is to use the `-V` option to `ctest` (and remove for now `--output-on-failure`) to see the output.

**Comment by v-dobrev** [comment]
> > Maybe because travis only allows 3 ranks?
> 
> This is not an issue in the GNU make build where we also use 4 processors.
> 
> To debug, one thing we can try is to use the `-V` option to `ctest` (and remove for now `--output-on-failure`) to see the output.

Actually, I was looking at the Mac tests -- that's where we use 4 processors for the MPI tests. On linux we actually use 2 processors, so we should probably do the same in the CMake build.

**Comment by v-dobrev** [comment]
The last failure on travis seems to be due to an issue with #1309.

**Comment by pazner** [comment]
Looks like the last commit (adding a `-h` help flag) interacts badly with the `runtest` script (which passes `mfem_dir` as a command line argument to the test).

I have a quick fix for this. Where's the best place to put it?

**Comment by v-dobrev** [comment]
I think it's OK to add it here -- this is where I noticed the isssue.

**Comment by pazner** [comment]
I created a new PR #1354 since this PR was already merged and the branch was deleted.

**Comment by pazner** [comment]
I created a new PR #1354 since #1309 was already merged and the branch was deleted. Feel free to grab the fix from there if you think it would be better in this PR.

**Comment by v-dobrev** [comment]
> I created a new PR #1354 since this PR was already merged and the branch was deleted.

Thanks, @pazner! This must be a record time between noticing a bug and creating a PR to fix it. 😄

**Comment by tzanio** [comment]
@pazner and @v-dobrev I meant to let you know yesterday that I noticed and fixed this issues in https://github.com/mfem/mfem/pull/1286/commits/424fd7e9faf866df9956dbd769bda66ba5ad7325. I like your fix more though, I'll merge it.

**Comment by pazner** [comment]
> @pazner and @v-dobrev I meant to let you know yesterday that I noticed and fixed this issues in [424fd7e](https://github.com/mfem/mfem/commit/424fd7e9faf866df9956dbd769bda66ba5ad7325). I like your fix more though, I'll merge it.

I'm working on making the command line arguments more robust. I should have a better solution soon.

**Comment by tzanio** [comment]
Ok, I'll wait

**Comment by v-dobrev** [comment]
I restarted the branch-history test on travis (after #1354 was merged) and it now looks good.

**Comment by tzanio** [comment]
@adrienbernede and @v-dobrev do you have any further questions, or can we merge this in `next`?

**Comment by tzanio** [comment]
Merged in `next` for testing...

**Pull request merged**